File: backend/app/sharing/service.py
import os
import re
from sqlalchemy.orm import Session
from app.models.user import User, RoleEnum
from .schemas import ShareProfileResponse

FRONTEND_URL = os.getenv("FRONTEND_URL", "http://localhost:5173")

class SharingService:
    def __init__(self, session: Session):
        self.session = session

    def get_share_info(self, user: User) -> ShareProfileResponse:
        base_url = FRONTEND_URL.rstrip("/")

        if user.role == RoleEnum.PROMOTER and user.promoter_profile:
            # Correct public route: /promoters/:username
            username = user.promoter_profile.username
            slug = username
            public_url = f"{base_url}/promoters/{slug}"

        elif user.role == RoleEnum.BUSINESS and user.business_profile:
            # Businesses don't have a public profile page — share their campaign marketplace listing
            company_name = user.business_profile.company_name
            slug = re.sub(r'[^a-z0-9]+', '-', company_name.lower()).strip('-') or str(user.id)[:8]
            username = company_name
            # Link to the campaign marketplace so people can find their campaigns
            public_url = f"{base_url}/promoter/marketplace"

        else:
            slug = str(user.id)[:8]
            username = user.full_name or "User"
            public_url = f"{base_url}/promoter/marketplace"

        return ShareProfileResponse(
            public_url=public_url,
            qr_code_url=None,  # Generated client-side via react-qr-code
            username=username,
            slug=slug
        )

File: backend/src/modules/sharing/service.js
import { config } from "../../config/env.js";
import { z } from "zod";
import { prisma } from "../../config/db.js";
import { AppError } from "../../shared/errors.js";

export const campaignShareParamsSchema = z.object({
  campaignId: z.string().min(1),
});

export function getProfileShare(user) {
  const base = config.frontendUrl;

  if (user.role === "PROMOTER" && user.promoterProfile) {
    const username = user.promoterProfile.username;
    return {
      publicUrl: `${base}/promoters/${username}`,
      qrCodeUrl: null,
      username,
      slug: username,
    };
  }

  if (user.role === "BUSINESS" && user.businessProfile) {
    const companyName = user.businessProfile.companyName;
    const slug =
      companyName
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, "-")
        .replace(/^-|-$/g, "") || user.id.slice(0, 8);
    return {
      publicUrl: `${base}/promoter/marketplace`,
      qrCodeUrl: null,
      username: companyName,
      slug,
    };
  }

  const slug = user.id.slice(0, 8);
  return {
    publicUrl: `${base}/promoter/marketplace`,
    qrCodeUrl: null,
    username: user.fullName || "User",
    slug,
  };
}

export async function getCampaignShare(user, campaignId) {
  const campaign = await prisma.campaign.findUnique({ where: { id: campaignId } });
  if (!campaign) throw new AppError("Campaign not found", 404);
  if (user.role !== "BUSINESS" || campaign.businessProfileId !== user.businessProfile?.id) {
    throw new AppError("You are not authorized to share this campaign", 403);
  }

  const base = config.frontendUrl;
  return {
    publicUrl: `${base}/campaign-marketplace/${campaignId}`,
    qrCodeUrl: null,
    campaignId,
    title: campaign.title,
  };
}
